pypdf/_cmap.py

```diff
@@ -423,57 +433,93 @@ def compute_space_width(
                 w = w[2:]
             else:
                 logger_warning(
                     "unknown widths : \n" + (ft1["/W"]).__repr__(),
                     __name__,
                 )
                 break
-        try:
-            sp_width = w1[space_code]
-        except Exception:
-            sp_width = (
-                w1[-1] / 2.0
-            )  # if using default we consider space will be only half size
     elif "/Widths" in ft:
-        w = list(ft["/Widths"])  # type: ignore
-        try:
-            st = cast(int, ft["/FirstChar"])
-            en: int = cast(int, ft["/LastChar"])
-            if st > space_code or en < space_code:
-                raise Exception("Not in range")
-            if w[space_code - st].get_object() == 0:
-                raise Exception("null width")
-            sp_width = w[space_code - st].get_object()
-        except Exception:
-            if "/FontDescriptor" in ft and "/MissingWidth" in cast(
-                DictionaryObject, ft["/FontDescriptor"]
-            ):
-                sp_width = ft["/FontDescriptor"]["/MissingWidth"].get_object()  # type: ignore
-            else:
-                # will consider width of char as avg(width)/2
-                m = 0
-                cpt = 0
-                for xx in w:
-                    xx = xx.get_object()
-                    if xx > 0:
-                        m += xx
-                        cpt += 1
-                sp_width = m / max(1, cpt) / 2
-
-    if is_null_or_none(sp_width):
-        sp_width = 0.0
+        w = list(ft["/Widths"]) if isinstance(ft["/Widths"], ArrayObject) else []
+        if "/FontDescriptor" in ft and "/MissingWidth" in cast(
+            DictionaryObject, ft["/FontDescriptor"]
+        ):
+            font_width_map["default"] = ft["/FontDescriptor"]["/MissingWidth"].get_object()  # type: ignore
+        else:
+            # will consider width of char as avg(width)
+            m = 0
+            cpt = 0
+            for xx in w:
+                xx = xx.get_object()
+                if xx > 0:
+                    m += xx
+                    cpt += 1
+            font_width_map["default"] = m / max(1, cpt)
+        st = cast(int, ft["/FirstChar"])
+        en = cast(int, ft["/LastChar"])
+        for c_code in range(st, en + 1):
+            try:
+                width = w[c_code - st].get_object()
+                font_width_map[c_code] = width
+            except Exception:
+                # The PDF structure is invalid. The array is too small
+                # for the specified font width.
+                pass
+    if is_null_or_none(font_width_map.get("default")):
+        font_width_map["default"] = default_font_width if default_font_width else 0.0
+    return font_width_map
+
+
+def compute_space_width(
+    font_width_map: Dict[Any, float], sp: int
+) -> float:
+    try:
+        sp_width = font_width_map[sp]
+        if sp_width == 0:
+            raise Exception("Zero width")
```

Comment:
This is too generic in my opinion.

Reply:
Sorry, I don't quite understand the nuances of English.

I think it would be fine to just input a value without using an Exception, but would that be preferable?

Do you mean that we should consider a more detailed code pattern?

Should I use ValueError or something else instead of raising with Exception?

Do you mean that we should revert to the original code name?

```
            if w[space_code - st].get_object() == 0:
                raise Exception("null width")
```

Reply:
I would prefer to work with more specific exceptions in general. This case is a bit special, as we are catching the exception directly afterwards, but a `ValueError` would make more sense in my opinion.

Reply:
Thank you!